### ISSUE TITLE
[Feature] Support alternate sources of Chunkserver identity

### ISSUE THREAD

**Issue by zicklag** [lizardfs/lizardfs#721]
Currently, when a LizardFS chunkserver joins to a cluster, the LizardFS master identifies the chunkserver based on its IP address. This creates issues when running LizardFS in containers as containers may get a [different IP address](https://github.com/lizardfs/lizardfs/issues/711#issuecomment-400762045) every time they restart.

It would be useful for LizardFS to support alternative methods of identifying chunkservers, such as using the hostname or an arbitrary name set in the chunkserver configuration. The default method of identifying chunkservers could still be based on the IP address, but you should be able to choose the method to be used in the master configuration.

**Comment by Blackpaw** [comment]
? Whats the problem with the chunkserver having a different IP each restart?

**Comment by zicklag** [comment]
@Blackpaw Because if a chunkserver container starts up with a different IP and therefore a different chunkserver id, the master thinks that it should have a different set of chunks on it and marks all of the chunks that used to be on it ( and actually still are ) as undergoal because it doesn't see a chunkserver that has them anymore. It just sees a new chunkserver that it doesn't have any information on.

**Comment by 4Dolio** [comment]
Really? I do not think that is what happens.
Yes, when a CS goes down it's chunks are missing.  When it boots up, even at a new IP, it scans and registers it's chunks with the master and everything should be stable again.  The master should not care what IP a CS uses, it will know the old IP is missing, and the mfs.cgi can click a link to forget it, but it's chunks were already lost from the master perspective, and then those chunks came back, at a new IP, so all is well... Right?

**Comment by Blackpaw** [comment]
Yup, just did a quick test with 3.13rc1
- Took a chunkserver down
- under goal by CS chunks
- Changed CS IP
- Brought it back up
- Chunks recognised, no longer under goal.

nb. For non trivial numbers of chunks it can take up to a few minutes for them to be scanned and integrated.

**Comment by Blackpaw** [comment]
Something I have noticed with DHCP and chunkservers - they don't like it. If the CS starts up before the network is up, I don't think the CS will try to reconnect.

**Comment by njhurst** [comment]
@Blackpaw I changed my target in the systemd .service file like this:
```
After=local-fs.target network-online.target lizardfs-master.service
Wants=network-online.target
```
Which solves the dhcp problem.

**Comment by Blackpaw** [comment]
Thanks njhurst. Works well with servers, but my problem was on a home media server that was using the network manager tray gui to start the connection (xfce), which doesn't work with "network-online.target".

I actually thought it was static :) hadn't checked it in years. Once I changed it to static, that and a bunch of other problems were resolved.

Getting OT though ...

**Comment by 4Dolio** [comment]
I am not saying something is not going wrong... but as you described your situation, I think it should be fine.  Lets keep looking for where the bug is.

**Comment by 4Dolio** [comment]
I am not saying something is not going wrong... but as you described your situation, I think it should be fine.  Lets keep looking for where the bug is.

**Comment by zicklag** [comment]
@Blackpaw 
> For non trivial numbers of chunks it can take up to a few minutes for them to be scanned and integrated.

That sounds right. It may not actually be a big enough problem, then, to warrant fixing. I had a bigger issue one time when the chunkservers had some problem and were restarting repeatedly which caused Gigabytes of data to get moved to the trash and fill up my relatively small drives and freeze the filesystem because it was out of space. It must not have had time to realize that the chunks on the servers were still there before the chunkserver restart yet again.

This was mostly to remedy my containerized solution for LizardFS (#711), but now I'm thinking that I might refactor the Docker plugin, anyway, in favor of a more stable solution.

I'm not sure if this would still be useful, or if it is not worth the trouble.

**Comment by Blackpaw** [comment]
>I'm not sure if this would still be useful, or if it is not worth the trouble.

I don't think it would change anything, the chunkservers are working as designed,

**Comment by mcassaniti** [comment]
I thought I might document here that setting a listen IP on the chunk server will cause that IP to be listed against the MFS master server. In particular I needed to set `CSSERV_LISTEN_HOST` to the IP that clients __and other chunk servers__ will connect to a chunk server on. My MFS master was running in a Docker swarm container (basically behind a NAT) and resulted in it seeing the wrong chunk server IP.

This does not work however if the chunk server is running in a container unless running with `--net host`. The listen IP specified will exist outside the container, but not within the container, resulting in the chunk server process being unable to bind the socket.

It would be nice to provide a chunk server advertise IP to the MFS server. The below should help elaborate. It would also be nice to set the hostname or some other name of the chunk server so that the name of the chunk server can be seen in the tools/CGI server. I realise that this requires extending the protocol.

| Role | Local IP | IP Seen By Master | IP to Connect to |
|-------|------------|-------------------------|----------------------|
| Master | 10.0.0.2 | N/A | 192.168.1.5 |
| Chunk Server | 172.19.0.2 | 10.255.0.3 | 192.168.1.3 |
